packages/data/src/hypoCompress.ts

```
@@ -0,0 +1,402 @@
+      return null;
+    }
+  } catch {
+    return null;
+  }
+
+  return compressed;
+}
+
+/**
+ * Decompresses a string into a JSON string, representing a hypo.
+ * Returns empty string if the decompression fails.
+ *
+ * @param compressedJSON A string representing a compressed hypo JSON
+ */
+export function expand(compressedJSON: string): string {
+  return gameJSONDecompress(compressedJSON);
+}
+
+/**
+ * Compresses a GameJSON object into a string.
+ * Returns empty string if decompression fails.
+ *
+ * @param data The GameJSON object
+ */
+function gameJSONCompress(data: GameJSON): string {
+  let out = "";
+
+  // number of players
+  const numberOfPlayers = data.players.length;
+  if (numberOfPlayers < 2 || numberOfPlayers > 6) {
+    return "";
+  }
+  out += `${data.players.length}`;
+
+  // deck
+  const [deck, okDeck] = compressDeck(data.deck);
+  if (!okDeck) {
+    return "";
+  }
+  out += `${deck},`;
+
+  // actions
+  const actions = compressActions(data.actions);
+  out += `${actions},`;
+
+  // variant id
+  const variant = variantsJSONArray.find(
+    (v) => v.name === data.options.variant,
+  );
+  if (variant === undefined) {
+    return "";
+  }
+  out += `${variant.id}`;
+
+  // add hyphens every 20 characters for URL posting (hyphens make the text wrap)
+  out = out.match(/.{1,20}/g)!.join("-");
+
+  return out;
+}
+
+/**
+ * Decompresses a string into a GameJSON object.
+ * Returns empty string if decompression fails.
+ *
+ * @param data The compressed string
+ */
+function gameJSONDecompress(data: string): string {
+  // remove all hyphens from URL
+  const normal = data.replace(/-/g, "");
+
+  // The compressed string is composed of 3 substrings, separated by comma.
+  const [deckString, actionsString, variantString] = [...normal.split(",", 3)];
+  const numPlayers = parseInt(deckString.charAt(0), 10);
+
+  const players = getPlayers(numPlayers);
+  if (players.length === 0) {
+    return "";
```

return null on failure instead of an empty string

---

marked as resolved but wasn't fixed